docs: Add Tags API endpoint documentation (#467) (#481)
mautic/developer-documentation-new: docs/rest_api/tags.rst

> Tags

Reviewer: Migrated content structure and API endpoint patterns from the legacy Tags API documentation.

Source: https://github.com/mautic/developer-documentation/blob/main/source/includes/_api_endpoint_tags.md

####

Use this endpoint to work with Mautic's Tags.

**Using Mautic's API Library**

> You can interact with this API through the :xref:`Mautic API Library` as follows, or use the various HTTP endpoints as described in this document.

Reviewer 2: ```suggestion
Using the Mautic API library
****************************

.. vale off

You can interact with this API using the :xref:`Mautic API Library` as below, or the various HTTP endpoints described in this document.

.. vale on
```

.. code-block:: php

   <?php
   use Mautic\MauticApi;
   use Mautic\Auth\ApiAuth;

   // ...
   $initAuth = new ApiAuth();
   $auth     = $initAuth->newAuth($settings);
   $apiUrl   = "https://example.com";
   $api      = new MauticApi();
   $tagApi   = $api->newApi("tags", $auth, $apiUrl);

.. vale off

Get Tag
*******

.. vale on

> 

Reviewer 2: Move the description here for consistency with other pages.

```suggestion

Retrieves an individual Tag.

```

.. code-block:: php

   <?php

   //...
   $tag = $tagApi->get($id);

.. code-block:: json

   {
     "tag": {
       "id": 34,
       "tag": "tagA",
       "description": "A tag for grouping contacts"
     }
   }

Retrieves an individual Tag.

> 

Reviewer 2: ```suggestion
```

.. vale off

> **HTTP Request**

Reviewer 2: ```suggestion
HTTP request
============
```

.. vale on

``GET /tags/ID``

> **Response**

Reviewer: Verified Tag entity fields against Mautic 7.0 source. Added `description` field (line 59) which is exposed via `loadApiMetadata()` (lines 81-92) but was missing from legacy docs.

Source: https://github.com/mautic/mautic/blob/7.0/app/bundles/LeadBundle/Entity/Tag.php

> **Response**

Reviewer 2: ```suggestion
Response
========
```

> ``Expected Response Code: 200``

Reviewer 2: ```suggestion
* Returns ``200 OK`` when the request successfully retrieves the Tag.
```

> See JSON code example.

Reviewer 2: There's no JSON code example here?

```suggestion
.. _get Tag response:

.. code-block:: json

   {  
       "tag": {
           "id": 34,
           "tag": "tagA",
       }
   }
```

> **Tag properties**

Reviewer 2: ```suggestion
.. _get Tag properties:

Tag properties
--------------
```

.. list-table::
   :header-rows: 1
   :widths: 25 15 60

   * - Name
     - Type
     - Description
   * - ``id``
     - int
     - ID of the Tag
   * - ``tag``
     - string
     - Title of the Tag
   * - ``description``
     - string/null
     - Description of the Tag

.. vale off

List Tags
*********

> 

Reviewer 2: ```suggestion

Retrieves a list of Tags.

```

.. vale on

.. code-block:: php

   <?php

   //...
   $tags = $tagApi->getList($searchFilter, $start, $limit, $orderBy, $orderByDir, $publishedOnly, $minimal);

.. code-block:: json

   {
     "total": 1,
     "tags": [
       {
         "id": 34,
         "tag": "tagA",
         "description": "A tag for grouping contacts"
       }
     ]
   }

> 

Reviewer 2: ```suggestion
```

> Retrieves a list of Tags.

Reviewer 2: ```suggestion
```

.. vale off

> **HTTP Request**

Reviewer 2: ```suggestion
HTTP request
============
```

.. vale on

``GET /tags``

> **Query parameters**

Reviewer 2: ```suggestion
Query parameters
----------------
```

.. list-table::
   :header-rows: 1
   :widths: 25 75

   * - Name
     - Description
   * - ``search``
     - String or search command to filter entities by.
   * - ``start``
     - Starting row for the entities returned. Defaults to 0.
   * - ``limit``
     - Limit number of entities to return. Defaults to the system configuration for pagination, which is 30 by default.
   * - ``orderBy``
     - Column to sort by. Can use any column listed in the response.
   * - ``orderByDir``
     - Sort direction: ``asc`` or ``desc``.
   * - ``publishedOnly``
     - Only return currently published entities.
   * - ``minimal``

>      - Return only array of entities without additional lists in it.

Reviewer 2: ```suggestion
.. list-table::
   :header-rows: 1
   :widths: 25 75

   * - Name
     - Description
   * - ``search``
     - String or search command to filter entities
   * - ``start``
     - Starting row for the returned entities - defaults to 0
   * - ``limit``
     - Maximum number of entities to return - defaults to 30
   * - ``orderBy``
     - Column to sort by. Any column in the response is valid.
   * - ``orderByDir``
     - Order direction - ``asc`` or ``desc``
   * - ``publishedOnly``
     - Returns only currently published entities
   * - ``minimal``
     - Returns only a simple mapped object of entities without additional lists in it
```

> **Response**

Reviewer 2: ```suggestion
Response
========
```

> ``Expected Response Code: 200``

Reviewer 2: ```suggestion
* Returns ``200 OK`` when the request successfully retrieves the Tags list.
```

> See JSON code example.

Reviewer 2: Move the json example here for consistency.

```suggestion
.. code-block:: json

   {
      "total": 1,
      "tags": [
         {
            "id": 34,
            "tag": "tagA",
            "description": "A tag for grouping contacts"
         }
      ]
   }
```

> **Properties**

Reviewer 2: ```suggestion
Properties
----------
```

> Same as `Get Tag <#get-tag>`_.

Reviewer 2: ```suggestion
.. list-table::
   :widths: 25 25 50
   :header-rows: 1

   * - Name
     - Type
     - Description
   * - ``total``
     - integer
     - Total count of Tags
   * - ``tags``
     - array
     - A mapped collection of Tags indexed by their ID

For the rest of the properties, refer to :ref:`Tag properties <get Tag properties>`.
```

.. vale off

Create Tag
**********

> 

Reviewer 2: ```suggestion

Creates a new Tag.

```

.. vale on

.. code-block:: php

   <?php

   $data = array(
       'tag' => 'Tag A',
       'description' => 'Description of Tag A',

>    );

Reviewer 3: Let's use more modern syntax.
```suggestion
   $data = [
       'tag' => 'Tag A',
       'description' => 'Description of Tag A',
   ;
```

   $tag = $tagApi->create($data);

> Creates a new Tag.

Reviewer 2: ```suggestion
```

.. vale off

> **HTTP Request**

Reviewer 2: ```suggestion
HTTP request
============
```

.. vale on

``POST /tags/new``

> **POST parameters**

Reviewer 2: ```suggestion
.. _create Tag POST parameters:

POST parameters
---------------
```

.. list-table::
   :header-rows: 1
   :widths: 25 15 60

   * - Name
     - Type
     - Description
   * - ``tag``
     - string

>      - Title of the Tag. Required.

Reviewer 2: ```suggestion
     - **Required.**
       
       Title of the Tag
```

   * - ``description``
     - string
     - Description of the Tag

> **Response**

Reviewer 2: ```suggestion
Response
========
```

> ``Expected Response Code: 201``

Reviewer 2: ```suggestion
* Returns ``201 Created`` when the request successfully creates an Tag.
```

> **Properties**

Reviewer 2: ```suggestion
Properties
----------
```

> Same as `Get Tag <#get-tag>`_.

Reviewer 2: ```suggestion
Refer to :ref:`Tag properties <get Tag properties>`.
```

.. vale off

Edit Tag
********

> 

Reviewer 2: ```suggestion

Edits a Tag.

This operation supports ``PUT`` or ``PATCH`` depending on the desired behavior:

* ``PUT``: **full replacement**. The request creates a new Tag if the ID doesn't exist. If the ID exists, the request clears all existing data and replaces it with the provided values.
* ``PATCH``: **partial update**. The request only updates field values based on the request data. The request fails when the Tag ID doesn't exist.

```

.. vale on

.. code-block:: php

   <?php

   $id   = 1;
   $data = array(
       'tag' => 'Tag B',
       'description' => 'Updated description',

>    );

Reviewer 3: ```suggestion
   $data = [
       'tag' => 'Tag B',
       'description' => 'Updated description',
   ];
```

   // Create new Tag if ID 1 isn't found?
   $createIfNotFound = true;

   $tag = $tagApi->edit($id, $data, $createIfNotFound);

Edits a Tag.

This operation supports ``PUT`` or ``PATCH`` depending on the desired behavior:

* ``PUT``: **full replacement**. The request creates a new Tag if the ID doesn't exist. If the ID exists, the request clears all existing data and replaces it with the provided values.
* ``PATCH``: **partial update**. The request only updates field values based on the request data. The request fails when the Tag ID doesn't exist.

> 

Reviewer 2: ```suggestion
```

.. vale off

> **HTTP Request**

Reviewer 2: ```suggestion
HTTP request
============
```

.. vale on

To edit a Tag and return a 404 if the Tag isn't found:

``PATCH /tags/ID/edit``

To edit a Tag and create a new one if the Tag isn't found:

> ``PUT /tags/ID/edit``

Reviewer 2: ```suggestion
* ``PUT /tags/ID/edit``: updates an existing Tag or creates a new one when the ID doesn't exist.
* ``PATCH /tags/ID/edit``: updates an existing Tag. The request fails when the ID doesn't exist.
```

> **POST parameters**

Reviewer 2: ```suggestion
POST parameters
---------------
```

.. list-table::
   :header-rows: 1
   :widths: 25 15 60

   * - Name
     - Type
     - Description
   * - ``tag``
     - string
     - Title of the Tag
   * - ``description``
     - string
     - Description of the Tag

**Response**

If using ``PUT``, the expected response code is ``200`` if editing the Tag or ``201`` if creating the Tag.

> If using ``PATCH``, the expected response code is ``200``.

Reviewer 2: ```suggestion
Response
========

* ``PUT``: returns ``200 OK`` when the request successfully updates the Tag or ``201 Created`` when the request creates a Tag.
* ``PATCH``: returns ``200 OK`` when the request successfully updates the Tag or ``404 Not Found`` error when the Tag ID doesn't exist.

The response is a JSON object similar to :ref:`Get Tag <get Tag response>`.
```

**Properties**

> Same as `Get Tag <#get-tag>`_.

Reviewer 2: ```suggestion
Properties
----------

Refer to :ref:`Tag properties <get Tag properties>`.
```

.. vale off

Delete Tag
**********

> 

Reviewer 2: ```suggestion

Deletes a Tag.

```

.. vale on

.. code-block:: php

   <?php

   $tag = $tagApi->delete($id);

Deletes a Tag.

> 

Reviewer 2: ```suggestion
```

.. vale off

> **HTTP Request**

Reviewer 2: ```suggestion
HTTP request
============
```

.. vale on

``DELETE /tags/ID/delete``

> **Response**

Reviewer 2: ```suggestion
Response
========
```

> ``Expected Response Code: 200``

Reviewer 2: ```suggestion
* Returns ``200 OK`` when the request successfully deletes the Tag.

The response is a JSON object containing the data of the deleted Tag, similar to :ref:`Get Tag <get Tag response>`.
```

> **Properties**

Reviewer 2: ```suggestion
Properties
----------
```

> Same as `Get Tag <#get-tag>`_.

Reviewer 2: ```suggestion
Refer to :ref:`Tag properties <get Tag properties>`.
```

* ``PUT``: **full replacement**. The request creates a new Tag if the ID doesn't exist. If the ID exists, the request clears all existing data and replaces it with the provided values.
* ``PATCH``: **partial update**. The request only updates field values based on the request data. The request fails when the Tag ID doesn't exist.

.. code-block:: php

   <?php

   $id   = 1;
   $data = [
       'tag' => 'Tag B',
       'description' => 'Updated description',
   ];

   // Create new Tag if ID 1 isn't found?
   $createIfNotFound = true;

   $tag = $tagApi->edit($id, $data, $createIfNotFound);

.. vale off

HTTP request
============

.. vale on

To edit a Tag and return a 404 if the Tag isn't found:

``PATCH /tags/ID/edit``

To edit a Tag and create a new one if the Tag isn't found:

``PUT /tags/ID/edit``

POST parameters
---------------

.. list-table::
   :header-rows: 1
   :widths: 25 15 60

   * - Name
     - Type
     - Description
   * - ``tag``
     - string
     - Title of the Tag
   * - ``description``
     - string
     - Description of the Tag

Response
========

If using ``PUT``, the expected response code is ``200`` if editing the Tag or ``201`` if creating the Tag.

> If using ``PATCH``, the expected response code is ``200``.

Reviewer 2: ```suggestion
* ``PUT``: returns ``200 OK`` when the request successfully updates the Tag or ``201 Created`` when the request creates a Tag.
* ``PATCH``: returns ``200 OK`` when the request successfully updates the Tag or ``404 Not Found`` error when the Tag ID doesn't exist.

The response is a JSON object similar to :ref:`Get Tag <get Tag response>`.
```

   $tagApi   = $api->newApi("tags", $auth, $apiUrl);

.. vale off

Get Tag
*******

.. vale on

Retrieves an individual Tag.

.. code-block:: php

   <?php

   //...
   $tag = $tagApi->get($id);

.. vale off

HTTP request
============

.. vale on

``GET /tags/ID``

Response
========

* Returns ``200 OK`` when the request successfully retrieves the Tag.

.. _get Tag response:

.. code-block:: json

   {
       "tag": {
           "id": 34,
           "tag": "tagA",
           "description": "A tag for grouping contacts"
       }
   }

.. _get Tag properties:

Tag properties
--------------

.. list-table::
   :header-rows: 1
   :widths: 25 15 60

   * - Name
     - Type
     - Description
   * - ``id``
     - int
     - ID of the Tag
   * - ``tag``
     - string
     - Title of the Tag
   * - ``description``
     - string/null
     - Description of the Tag

.. vale off

List Tags
*********

.. vale on

Retrieves a list of Tags.

.. code-block:: php

   <?php

   //...
   $tags = $tagApi->getList($searchFilter, $start, $limit, $orderBy, $orderByDir, $publishedOnly, $minimal);

.. vale off

HTTP request
============

.. vale on

``GET /tags``

Query parameters
----------------

> .. list-table::
>    :header-rows: 1
>    :widths: 25 75
> 
>    * - Name
>      - Description
>    * - ``search``
>      - String or search command to filter entities
>    * - ``start``
>      - Starting row for the returned entities - defaults to 0
>    * - ``limit``
>      - Maximum number of entities to return - defaults to 30
>    * - ``orderBy``
>      - Column to sort by. Any column in the response is valid.
>    * - ``orderByDir``
>      - Order direction - ``asc`` or ``desc``
>    * - ``publishedOnly``
>      - Returns only currently published entities
>    * - ``minimal``
>      - Returns only a simple mapped object of entities without additional lists in it

Reviewer 2: Let's wrap the list table in vale on/off statements.

```suggestion
.. vale off

.. list-table::
   :header-rows: 1
   :widths: 25 75

   * - Name
     - Description
   * - ``search``
     - String or search command to filter entities
   * - ``start``
     - Starting row for the returned entities - defaults to 0
   * - ``limit``
     - Maximum number of entities to return - defaults to 30
   * - ``orderBy``
     - Column to sort by. Any column in the response is valid.
   * - ``orderByDir``
     - Order direction - ``asc`` or ``desc``
   * - ``publishedOnly``
     - Returns only currently published entities
   * - ``minimal``
     - Returns only a simple mapped object of entities without additional lists in it

.. vale on
```

Response
========

* Returns ``200 OK`` when the request successfully retrieves the Tags list.

.. code-block:: json

   {
       "total": 1,
       "tags": [
           {
               "id": 34,
               "tag": "tagA",
               "description": "A tag for grouping contacts"
           }
       ]
   }

Properties
----------

.. list-table::
   :widths: 25 25 50
   :header-rows: 1

   * - Name
     - Type
     - Description
   * - ``total``
     - integer
     - Total count of Tags
   * - ``tags``
     - array
     - A mapped collection of Tags indexed by their ID

> For the rest of the properties, refer to :ref:`Tag properties <get Tag properties>`.

Reviewer 2: ```suggestion
.. vale off

For the rest of the properties, refer to :ref:`Tag properties <get Tag properties>`.

.. vale on
```

Tags
####

> Use this endpoint to work with Mautic's Tags.

Reviewer 2: @promptless-for-oss please address the suggestion below for consistency.

```suggestion
Use this endpoint to manipulate and obtain details on Mautic’s Tags.
```
